### PR TITLE
PUBDEV-8381: Add Gains/Lift table to Performance & Prediction

### PR DESCRIPTION
**Pull request by hannah-tillman**
For: [PUBDEV-8381](https://h2oai.atlassian.net/browse/PUBDEV-8381)

First draft of gains/lift added to performance and prediction. It's an amalgam of information from gains/lift in the flow documentation and that wonderful doc @wendycwong curated. 

Please let me know what needs to be changed/clarified/updated!

### CODE REVIEW

File: h2o-docs/src/product/performance-and-prediction.rst [hunk 1]
```diff
@@ -2004,6 +2004,108 @@ Examples:
         # build the standardized coefficient magnitudes plot:
         glm.std_coef_plot()
 
+
+Gains/Lift 
+''''''''''
+
+Gains/Lift evaluates the prediction ability of a binary classification model. The chart is computed using the prediction probability and the true response labels. The accuracy of the classification model for a random sample is evaluated according to the results when the model is and is not used. The Gains/Lift chart shows the effectiveness of the current model(s) compared to a baseline, allowing users to quickly identify the most useful model. 
+
```
Comment (wendycwong) [inline h2o-docs/src/product/performance-and-prediction.rst:2012]: Please change:

The accuracy of the classification model for a random sample is evaluated according to the results when the model is and is not used.

to

The accuracy of the baseline is evaluated when no model is used.  For instance, if there are x% positive responses in the dataset, when we grab 10% of the dataset, we can assume that there are 10% of the x% positive responses in the 10% dataset we chose.  If x = 10 and there are 10% positive responses in the dataset.  When we choose a 10% dataset, we will expect there to be 1% positive responses in the 10% dataset we chose.

File: h2o-docs/src/product/performance-and-prediction.rst [hunk 2]
```diff
@@ -2004,6 +2004,108 @@ Examples:
...
+To compute Gains/Lift, H2O applies the model to the original dataset to find the response probability. The data is divided into groups by quantile thresholds of the response probability. The default number of groups is 16; if there are fewer than sixteen unique probability values, then the number of groups is reduced to the number of unique quantile thresholds. 
+
+**An example**: a response model predicts who will respond to a marketing campaign. If you have a response model, you can make more detailed predictions. You use the response model to assign a score to all 100000 customers and predict the results of contacting only the top 10000 customers, the top 20000 customers, and so on. You do this by:
+
+- taking the dataset, sending it through your model, and obtaining a list of predicted output which is the probability of positive response;
+- sorting your dataset according to the output of your model which is the probability of positive response (this probability can also be called the **score**)
+
```
Comment (wendycwong) [inline h2o-docs/src/product/performance-and-prediction.rst:2019]: Please change 

sorting your dataset according to the output of your model which is the probability of positive response (this probability can also be called the **score**)

to

sorting your dataset according to the output of your model which is the probability of positive response (this probability can also be called the **score**) from highest to lowest;

File: h2o-docs/src/product/performance-and-prediction.rst [hunk 3]
```diff
@@ -2004,6 +2004,108 @@ Examples:
...
+    :scale: 30%
+
+In addition to the chart, a **Gains/Lift table** is also available. This table reports the following for each group:
+
+- **Cumulative data fractions**: fraction of data used to calculate gain and lift at
+- **Lower threshold**: the lowest score output of the dataset in the data fraction bin
+- **Average response rate**: total number of positive classes / total number of data rows in the dataset
+- **Lift and Cumulative lift**: ratio of cumulative response rate and average response rate.
+- **Score**: the average of all the classifier output probabilities for each individual data fraction bin
+- **Cumulative response rate**: (based on the above written example) for the first bin, it is the number of positive response over 10000; for the second bin, it is the ratio of the sum of positive response over the first and second bins and 20000; for the third bin, it is the ratio of the sum of positive response over the first, second, and third bins and 30000
```
Comment (wendycwong) [inline h2o-docs/src/product/performance-and-prediction.rst:2055]: please add after 10000:

(assume each bin contains 10000 rows)

**Review by wendycwong — CHANGES_REQUESTED** [review]
Thanks Hannah, almost there.

**Comment by ledell** [comment]
@hannah-tillman If you can make @wendycwong's additional changes, I will do a final review.  Thanks!

**Comment by hannah-tillman** [comment]
> @hannah-tillman If you can make @wendycwong's additional changes, I will do a final review. Thanks!

Changes are completed -- @ledell ready when you are :)

**Comment by michalkurka** [comment]
@maurever as an uplift expert - please give us your comments

File: h2o-docs/src/product/performance-and-prediction.rst [hunk 4]
```diff
@@ -2004,6 +2004,111 @@ Examples:
...
+.. figure:: images/gainslift_plot.png
+    :alt: Gains/Lift Plot
+    :scale: 30%
+
+In addition to the chart, a **Gains/Lift table** is also available. This table reports the following for each group:
+
+- **Cumulative data fractions**: fraction of data used to calculate gain and lift at
+- **Lower threshold**: the lowest score output of the dataset in the data fraction bin
+- **Average response rate**: total number of positive classes / total number of data rows in the dataset
+- **Lift and Cumulative lift**: ratio of cumulative response rate and average response rate.
```
Comment (maurever) [inline h2o-docs/src/product/performance-and-prediction.rst:2056]:
```suggestion
- **Lift**: ratio of response rate of the current data fraction and average response rate
- **Cumulative lift**: ratio of cumulative response rate and average response rate
```

File: h2o-docs/src/product/performance-and-prediction.rst [hunk 5]
```diff
@@ -2004,6 +2004,111 @@ Examples:
...
+- **Score**: the average of all the classifier output probabilities for each individual data fraction bin
+- **Cumulative response rate**: for the first bin, it is the number of positive response over 10000 (assume each bin contains 100000 rows); for the second bin, it is the ratio of the sum of positive response over the first and second bins and 20000; for the third bin, it is the ratio of the sum of positive response over the first, second, and third bins and 30000
+- **Capture rate**: for each bin it is the ratio of positive classes in each bin divided by the total number of positive classes in your dataset
+- **Cumulative capture rate**: for the first bin, it is just the capture rate; the second bin is the ratio of (sum of number of positive classes for the first two bins) and the total number of positive classes in your dataset
+- `Kolmogorov-Smirnov (KS) Metric`_
+- **Gain and Cumulative gain**: calculated for each bin in a cumulative manner:
```
Comment (maurever) [inline h2o-docs/src/product/performance-and-prediction.rst:2062]: Again I would explain the gain and cumulative gain separately. Also from which source do you take the definition? From code, I see a different definition:

https://github.com/h2oai/h2o-3/blob/a69892d34bf66ec5831be79fd7af93574041a296/h2o-core/src/main/java/hex/GainsLift.java#L185-L186

You can use `lift` instead of `number of positive responses in first bin / number of data in first bin`. The result value is then multiplied by 100.

So my suggestions:

```suggestion
- **Gain**:  :math: 100 \times (lift for current data fraction -1)  
- **Cumulative Gain**:  :math: 100 \times (cumulative lift for current data fraction -1)  
```

Then the table below is not necessary. Or you can use this table at the beginning to explain how cumulative counts work in general. It is the same for cumulative response rate, cumulative score, cumulative capture rate, etc.

**Review by maurever — CHANGES_REQUESTED** [review]

**Review by maurever — APPROVED** [review]
Looks good to me, thanks @hannah-tillman!

**Review by wendycwong — APPROVED** [review]

**Comment by ledell** [comment]
@hannah-tillman Can you add the Python version here?  There's an example in this ticket: https://h2oai.atlassian.net/browse/PUBDEV-8388

Screen Shot 2021-11-11 at 12 26 06 PM

File: h2o-docs/src/product/performance-and-prediction.rst [hunk 6]
```diff
@@ -2004,6 +2004,98 @@ Examples:
...
+
+        # Build and train the model:
+         model <- h2o.gbm(x = c("Origin", "Distance"), 
+                          y = "IsDepDelayed", 
+                          training_frame = airlines, 
+                          ntrees = 1, 
+                          gainslift_bins = 20)
+
+         # Plot the Gains/Lift chart:
+         gain_table = model@model$training_metrics@metrics$gains_lift_table
```
Comment (ledell) [inline h2o-docs/src/product/performance-and-prediction.rst:2042]:
```suggestion
         gl <- model@model$training_metrics@metrics$gains_lift_table
```

File: h2o-docs/src/product/performance-and-prediction.rst [hunk 7]
```diff
@@ -2004,6 +2004,98 @@ Examples:
...
+         plot(gain_table$cumulative_data_fraction,
```
Comment (ledell) [inline h2o-docs/src/product/performance-and-prediction.rst:2043]: R code should be changed to:

```r
plot(gl$cumulative_data_fraction,
     gl$cumulative_capture_rate,'l', 
     ylim = c(0,2.0), col = "darkgreen",
     xlab = "cumulative_data_fraction",
     ylab = "cumulative_capture_rate, cumulative_lift",
     main = "Gains/Lift")
lines(gl$cumulative_data_fraction,
      gl$cumulative_lift)
```

Comment (ledell) [reply h2o-docs/src/product/performance-and-prediction.rst:2043]: And we need a similar thing in Python.  It should look like the Flow plot that's in this ticket: https://h2oai.atlassian.net/browse/PUBDEV-8388

File: h2o-docs/src/product/performance-and-prediction.rst [hunk 8]
```diff
@@ -2004,6 +2004,98 @@ Examples:
...
+              gain_table$cumulative_capture_rate,'l')
+
+.. figure:: images/gainslift_plot.png
```
Comment (ledell) [inline h2o-docs/src/product/performance-and-prediction.rst:2046]: Plot needs to be updated with image like this (or you can use this one):

Screen Shot 2021-11-11 at 1 05 14 PM

**Review by ledell — CHANGES_REQUESTED** [review]
@maurever @wendycwong @hannah-tillman I had some feedback about the plots... In Flow, we show Gains/Lift on the same plot.  Is that what we want to do here?  I assumed so, so I made some changes.  

However, now I am not sure if that's what we want to do -- do we want to have separate plots for Gains and Lift?

Some other tools like sklearn have [separate plots](https://coderzcolumn.com/tutorials/machine-learning/scikit-plot-visualizing-machine-learning-algorithm-results-and-performance#3.7) for both.  I assumed we want to replicate what's in Flow, but now I am not sure.

File: h2o-docs/src/product/performance-and-prediction.rst [hunk 9]
```diff
@@ -2004,6 +2004,98 @@ Examples:
...
+
+To compute Gains/Lift, H2O applies the model to the original dataset to find the response probability. The data is divided into groups by quantile thresholds of the response probability. The default number of groups is 16; if there are fewer than sixteen unique probability values, then the number of groups is reduced to the number of unique quantile thresholds. 
+
+**An example**: a response model predicts who will respond to a marketing campaign. If you have a response model, you can make more detailed predictions. You use the response model to assign a score to all 100000 customers and predict the results of contacting only the top 10000 customers, the top 20000 customers, and so on. You do this by:
```
Comment (preranapradhan25) [inline h2o-docs/src/product/performance-and-prediction.rst:2017]: A small suggestion: 
In large numbers, adding commas improve readability.
You use the response model to assign a score to all **100,000** customers and predict the results of contacting only the top 10000 customers, the top **20,000** customers, and so on.

File: h2o-docs/src/product/performance-and-prediction.rst [hunk 10]
```diff
@@ -2004,6 +2004,98 @@ Examples:
...
+
+- taking the dataset, sending it through your model, and obtaining a list of predicted output which is the probability of positive response;
+- sorting your dataset according to the output of your model which is the probability of positive response (this probability can also be called the **score**) from highest to lowest;
+    
+    - In this case, the first bin contains the top 10000 customers with the highest response probability, the second bin contains the next 10000 customers with the highest response probability, and so on.
```
Comment (preranapradhan25) [inline h2o-docs/src/product/performance-and-prediction.rst:2022]: Same comment as above: In large numbers, adding commas improve readability.

**Comment by wendycwong** [comment]
@ledell Regarding the plots, usually I see gain/lift charts separate like these:

image

**Comment by wendycwong** [comment]
Either one will work for me.

**Comment by maurever** [comment]
@ledell @wendycwong 

I am also wondering, why we call it Gains/Lift, other implementation just call it Gain/Lift. Also in an uplift libraries, the name is just Gain, without _s_. :thinking: 

In uplift libraries the plots are usually also separate (CausalML, UplftML).

**Comment by michalkurka** [comment]
@ledell I  noticed you rejected the PR with this note:

>  @maurever @wendycwong @hannah-tillman I had some feedback about the plots... In Flow, we show Gains/Lift on the same plot. Is that what we want to do here? I assumed so, so I made some changes.

> However, now I am not sure if that's what we want to do -- do we want to have separate plots for Gains and Lift?

> Some other tools like sklearn have separate plots for both. I assumed we want to replicate what's in Flow, but now I am not sure.

It is not clear to me whether you are objecting to how it is done in H2O or the documentation.

In this PR please lets only focus on the documentation part.

**Comment by ledell** [comment]
@michalkurka The reason I rejected the PR is that the Python code snippet is currently missing and the image that's provided is not correct.  It's supposed to be Gains/Lift but only showed Gains.  So there's a number of small errors in the documentation itself.  I don't have issues with the implementation, just the documentation.

Currently, we don't have an "implementation" of the plots... it's shown how to do them from scratch in the docs.  So there's a [ticket open](https://h2oai.atlassian.net/browse/PUBDEV-8388) for that.  We currently were aiming to replicate the both-on-one-plot conecpt that we already have in Flow, but i am seeing a lot of examples from other tools where they plot them separately, so we could add a flag to the proposed `h2o.plot_gainslift()` function to plot one at a time, or make an additional function if we wanted to.

**Comment by michalkurka** [comment]
> @michalkurka The reason I rejected the PR is that the Python code snippet is currently missing and the image that's provided is not correct. It's supposed to be Gains/Lift but only showed Gains. So there's a number of small errors in the documentation itself. I don't have issues with the implementation, just the documentation.

Thanks for the clarification. @hannah-tillman please incorporate the feedback.

For everyone - in this PR lets please focus on the documentation changes only. We can brainstorm elsewhere. It is getting hard to read.

**Comment by wendycwong** [comment]
@hannah-tillman @ledell Use this one?
image

File: h2o-docs/src/product/performance-and-prediction.rst [hunk 11]
```diff
@@ -2004,6 +2004,144 @@ Examples:
...
+
+        # Build and train the model:
+        model = H2OGradientBoostingEstimator(ntrees=1, gainslift_bins=20)
+        model.train(x=["Origin","Distance"], 
+                    y="IsDepDelayed", 
+                    training_frame=airlines)
+
+        # Plot the Gains/Lift chart:
+        gl = model.gains_lift()
+
```
Comment (ledell) [inline h2o-docs/src/product/performance-and-prediction.rst:2069]:
```suggestion
```

**Review by ledell — APPROVED** [review]
Looks great!  Thank you so much @hannah-tillman @wendycwong!
